lib/src/wallet.rs

```diff
@@ -773,28 +774,51 @@ impl LightWallet {
         }
     }
 
-    pub async fn zbalance(&self, addr: Option<String>) -> u64 {
+    pub async fn sapling_balance(&self, addr: Option<String>) -> u64 {
+        self.shielded_balance::<SaplingNoteAndMetadata>(addr, &[])
+            .await
+    }
+
+    pub async fn orchard_balance(&self, addr: Option<String>) -> u64 {
+        self.shielded_balance::<OrchardNoteAndMetadata>(addr, &[])
+            .await
+    }
+
+    async fn shielded_balance<NnMd>(
+        &self,
+        addr: Option<String>,
+        filters: &[Box<dyn Fn(&&NnMd, &TransactionMetadata) -> bool + '_>],
+    ) -> u64
+    where
+        NnMd: traits::NoteAndMetadata,
+    {
         self.transactions
             .read()
             .await
             .current
             .values()
             .map(|transaction| {
-                transaction
-                    .sapling_notes
-                    .iter()
-                    .filter(|nd| match addr.as_ref() {
-                        Some(a) => {
-                            *a == encode_payment_address(
-                                self.config.hrp_sapling_address(),
-                                &nd.extfvk.fvk.vk.to_payment_address(nd.diversifier).unwrap(),
-                            )
-                        }
-                        None => true,
-                    })
+                let mut filtered_notes: Box<dyn Iterator<Item = &NnMd>> = Box::new(
+                    NnMd::transaction_metadata_notes(transaction)
+                        .iter()
+                        .filter(|nd| match addr.as_ref() {
+                            Some(a) => {
+                                use self::traits::Recipient as _;
+                                let diversified_address =
+                                    &nd.fvk().diversified_address(*nd.diversifier()).unwrap();
+                                *a == diversified_address.b32encode_for_network(&self.config.chain)
+                            }
+                            None => true,
```

Comment:
Should this be `false`?

Reply:
If we supply an address we're asking for the balance of that address. If we don't supply an address, we're asking for the balance of all addresses.